aioftp/server.py

```diff
@@ -452,6 +452,7 @@ class AbstractServer:
         self.server.close()
         tasks = [self.server.wait_closed()]
         for connection in self.connections.values():
+            connection.command_connection.close()
```

Comment:
Actually, in test under linux, the connection hangs on after close server. If I add connection.close(), the connection will close and stop server immediately

Reply:
Please add test with "hang on" behavior.

Reply:
It doesn't always occupy while testing. But add that line, the server can always close.
If there is no connection, the server can always close too.
I don't mind if you refuse this pr, I can use it by myself.
Thanks!

Reply:
> It looks like this is not necessary, since command connection closed in `finally` part of dispatcher. https://github.com/aio-libs/aioftp/blob/master/aioftp/server.py#L949

Maybe ssl problem or exception before stream.close() ?
I really don't know where is the problem, but first close() then dispatcher.cancel() is usefull for me.

Reply:
I suggest you provide some code to reproduce "hang on", so I can review the problem. If you can't, then I need some logs for "hang on" case.

Reply:
import threading
import os
import asyncio, ssl
import aioftp
class FtpServer(threading.Thread):
	def __init__(self):
		threading.Thread.__init__(self)
		self.setDaemon(True)
	def run(self,*args,**kwargs):
		self.loop = asyncio.new_event_loop()
		asyncio.set_event_loop(self.loop)
		ssl_server = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
		ssl_server.verify_mode = ssl.CERT_NONE
		ssl_server.check_hostname = False
		ssl_server.load_cert_chain(os.path.dirname(__file__) + os.sep + 'ssl.crt', os.path.dirname(__file__) + os.sep + 'ssl.key')
		self.server = aioftp.Server(
			aioftp.MemoryUserManager(None),
			socket_timeout=60,
			idle_timeout=60,
			data_ports=range(30000,30010),
			ssl=ssl_server,
		)
		self.loop.run_until_complete(self.server.start(None, 990))
		self.loop.run_forever()
		
	async def terminateserver(self):
		await self.server.close()
		self.loop.call_soon_threadsafe(self.loop.stop)

	def stop(self):
		asyncio.run_coroutine_threadsafe(self.terminateserver(), self.loop)

ftp = FtpServer()
ftp.start()
while True:
	print('[ s ]: Stop')
	print('[ e ]: Exit')
	try:
		cmd = input('Input: \r\n')
	except (KeyboardInterrupt, SystemExit):
		cmd = 'e'
	if cmd == 's':
		ftp.close()
	elif cmd == 'e':
		break

Reply:
Can't reproduce hang on with your pull but commented `connection.command_connection.close()`
``` python
import asyncio
import ssl
import aioftp
import logging


ssl_server = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
ssl_server.verify_mode = ssl.CERT_NONE
ssl_server.check_hostname = False
ssl_server.load_cert_chain("ssl.crt", "ssl.key")
server = aioftp.Server(
    aioftp.MemoryUserManager(None),
    socket_timeout=60,
    idle_timeout=60,
    data_ports=range(30000, 30010),
    ssl=ssl_server,
)
logging.basicConfig(level=logging.DEBUG)
loop = asyncio.get_event_loop()
loop.run_until_complete(server.start(None, 8021))
try:
    loop.run_forever()
except KeyboardInterrupt:
    loop.run_until_complete(server.close())
```
What ftps client do you use? I test server with `lftp` with sequence:
``` bash
$ lftp ftps://localhost:8021 
lftp localhost:~> ls
# ls output
```
And have this logging info:
``` bash
INFO:aioftp.server:serving on :::8021
INFO:aioftp.server:serving on 0.0.0.0:8021
INFO:aioftp.server:new connection from 127.0.0.1:43784
INFO:aioftp.server:220 welcome
INFO:aioftp.server:FEAT
INFO:aioftp.server:502 'feat' not implemented
INFO:aioftp.server:USER anonymous
INFO:aioftp.server:230 anonymous login
INFO:aioftp.server:PWD
INFO:aioftp.server:257 "/"
INFO:aioftp.server:PBSZ 0
INFO:aioftp.server:200 
INFO:aioftp.server:PROT P
INFO:aioftp.server:200 
INFO:aioftp.server:PASV
INFO:aioftp.server:227-listen socket created
INFO:aioftp.server:227 (127,0,0,1,117,48)
INFO:aioftp.server:LIST
INFO:aioftp.server:150 list transfer started
INFO:aioftp.server:226 list transfer done
# here I hit ctrl-c
INFO:aioftp.server:waiting for 2 tasks
INFO:aioftp.server:closing connection from 127.0.0.1:43784
```
Any thing I missed?

Reply:
Don't use ctrl+c, it will send TERM signal to force stop.
Use buildin function server.close()

Reply:
And my client is FileZilla

Reply:
My env is stoping ftp server without stop the process, and aioftp runs in a seperated thread.

Reply:
I forgot to mention that to keep user uploading large file away from interrupt by idletimeout, I set idletimeout to a huge number.
Maybe waiting task for long time?

And I don't know this is  a bug or not? User uploads a file exceed idletimeout will be force close data connection.

Reply:
> Don't use ctrl+c, it will send TERM signal to force stop.

Actually it will not, try this code with ctrl-c:
``` python
import time
while True:
    try:
        time.sleep(10 ** 5)
    except KeyboardInterrupt:
        print("nope")
```
But with sigterm signal from htop it will be terminated. Also, you can see proper task cancelation mechanic work in log messages, which closes connection and do not hang on.

> Use buildin function server.close()

It is called in exception handler

Reply:
> I forgot to mention that to keep user uploading large file away from interrupt by idletimeout, I set idletimeout to a huge number.
> Maybe waiting task for long time?
> 
> And I don't know this is a bug or not? User uploads a file exceed idletimeout will be force close data connection.

By default idletimeout is `None`, which means no timeout. It is assumed, that if you set idletimeout, then you known what you are doing. And yes, this looks like normal behavior.

Reply:
Ok, even your code do not "hang on" in my test case with `lftp`. I have two suggestions:
* Enable logging with code below and paste console output for "hang on" case
    ``` python
    import logging
    logging.basicConfig(level=logging.DEBUG)
    ```
* Provide exact step by step instruction to reproduce "hang on".